Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/overlay/InputOverlay.java

```diff
@@ -807,61 +812,48 @@ else if (mPreferences.getInt("wiiController", 3) == 4)
     refreshControls();
   }
 
-  private void saveControlPosition(int sharedPrefsId, int x, int y, int controller,
+  private void saveControlPosition(int buttonId, int x, int y, int controller,
           String orientation)
   {
-    final SharedPreferences sPrefs = PreferenceManager.getDefaultSharedPreferences(getContext());
-    SharedPreferences.Editor sPrefsEditor = sPrefs.edit();
-    sPrefsEditor.putFloat(getXKey(sharedPrefsId, controller, orientation), x);
-    sPrefsEditor.putFloat(getYKey(sharedPrefsId, controller, orientation), y);
-    sPrefsEditor.apply();
+    FloatSetting.valueOf(getXKey(buttonId, controller, orientation)).setFloat(mSettings, x);
+    FloatSetting.valueOf(getYKey(buttonId, controller, orientation)).setFloat(mSettings, y);
```

Comment:
It seems like we're only ever storing and loading ints with these settings even though they're floats. If we're going to do it like that, the settings should probably be IntSettings instead, but... What do you think about taking the opportunity to change the ways these settings are stored? If we store a float from 0 to 1 (where 0 is one edge of the screen and 1 is the other) instead of storing a number of pixels, you won't have the problem of controls ending up outside of the screen if you move settings between different devices or if you switch between split-screen mode and regular mode. (Though, getting it working properly in split-screen mode would also require us to stop using `DisplayMetrics`, which should probably be done in a separate PR so as to not have too much stuff in one PR.)

Reply:
Also, please call `saveSettings` after the user has changed something.

Reply:
I like the idea of changing the range from 0 to 1 but not in this PR. I made that a TODO. Bounds checking will be required for a change like that anyway so this still should be good to merge.

And I'm still confused about when `saveSettings` needs to be called. The modified value gets saved to the INI without calling `saveSettings`. I assume it's being saved somewhere in the C++ side. Why is calling `saveSettings` needed here?

Reply:
> I like the idea of changing the range from 0 to 1 but not in this PR. I made that a TODO. Bounds checking will be required for a change like that anyway so this still should be good to merge.

If we merge this as-is, we will need to add migration code for ignoring or rescaling old values once we change it to go from 0 to 1. That's the reason I brought this up now. I don't believe we would need any kind of bounds checking if we make the change at the same time as we switch the positions over from SharedPreferences to the INI file.

> And I'm still confused about when `saveSettings` needs to be called. The modified value gets saved to the INI without calling `saveSettings`. I assume it's being saved somewhere in the C++ side. Why is calling `saveSettings` needed here?

I think I forgot that we added a call to `saveSettings` in `EmulationActivity.onSaveInstanceState`. With that call there, we don't need an extra `saveSettings` call here. But otherwise we would've needed one.

(`saveSettings` needs to be called at least once before the app process gets killed, but it doesn't have to be called on the same `Settings` instance as the one that was used for actually making the changes. Well, as long as legacy settings aren't involved, that is...)

Reply:
> If we merge this as-is, we will need to add migration code for ignoring or rescaling old values once we change it to go from 0 to 1. That's the reason I brought this up now. I don't believe we would need any kind of bounds checking if we make the change at the same time as we switch the positions over from SharedPreferences to the INI file.

Shouldn't we have bounds checking anyway to prevent users from manually setting invalid values after switching to a 0 to 1 range? These could simply be treated as invalid values when that time comes.

> I think I forgot that we added a call to saveSettings in EmulationActivity.onSaveInstanceState. With that call there, we don't need an extra saveSettings call here. But otherwise we would've needed one.

I assumed the same thing too but I put a breakpoint in `saveSettings` and it wasn't triggered when starting emulation, changing a button position, then exiting emulation. The setting still was still saved to the INI though.

Reply:
> Shouldn't we have bounds checking anyway to prevent users from manually setting invalid values after switching to a 0 to 1 range? These could simply be treated as invalid values when that time comes.

Well, if the user manually puts buttons outside of the screen, it's their own fault that the buttons are now outside of the screen :) This isn't one of those situations where invalid values cause crashes, and it's always possible to solve the problem by resetting the button positions. But if you want to add bounds checks anyway, I'm not going to object to it.

> I assumed the same thing too but I put a breakpoint in `saveSettings` and it wasn't triggered when starting emulation, changing a button position, then exiting emulation. The setting still was still saved to the INI though.

I actually realized after leaving my comment that the Java `saveSettings` actually isn't getting called on emulation exit the way things are right now - see PR #9230. I'm assuming that what happened in your case is that there was no call to the Java `saveSettings` but there was a call to the C++ function which `saveSettings` calls. I'm not actually sure what C++ code would've done that, though. Probably something that runs when exiting emulation...

Reply:
> Well, if the user manually puts buttons outside of the screen, it's their own fault that the buttons are now outside of the screen :) This isn't one of those situations where invalid values cause crashes, and it's always possible to solve the problem by resetting the button positions. But if you want to add bounds checks anyway, I'm not going to object to it.

I was thinking of bounds checking similar to the latest commit. It doesn't cover every case but it should be good enough. Is this acceptable?